File: Blueberry.mx/spec/controllers/mixes_controller_spec.rb
require 'rails_helper'

describe MixesController do
  render_views
  describe "index" do
    before do
      Mix.create!(name: 'Supersonic Overdrive')
      Mix.create!(name: 'ChilZone')
      Mix.create!(name: 'Pot Pourri')

      xhr :get, :index, format: :json, keywords: keywords
    end
    
    subject(:results) { JSON.parse(response.body) }

    def extract_name
      ->(object) { object["name"] }
    end
  

    context "when the search finds results" do
      let(:keywords) { 'on' }
      
      it 'should 200' do
        expect(response.status).to eq(200)
      end

      it 'should return 2 results' do
        expect(results.size).to eq(2)
      end

      it 'should include "Supersonic Overdrive"' do
        expect(results.map(&extract_name)).to include('Supersonic Overdrive')
      end

      it 'should include "ChilZone"' do
        expect(results.map(&extract_name)).to include('ChilZone')
      end
    end

    context "when the search doesn't find results" do
      let(:keywords) { 'foo' }
      it 'should return no results' do
        expect(results.size).to eq(0)
      end
    end

    
  end
end

File: Blueberry.mx/spec/javascripts/controllers/MixesController_spec.js
describe("MixesController", function() {
  var ctrl, location, resource, routeParams, scope, setupController, httpBackend;

  setupController = function(keywords, results) {
    return inject(function($location, $routeParams, $rootScope, $resource, $httpBackend, $controller) {
      scope       = $rootScope.$new();
      location    = $location;
      resource    = $resource;
      routeParams = $routeParams;
      httpBackend = $httpBackend;

      routeParams.keywords = keywords;

      if (results) {
        request = new RegExp("\/mixList.*keywords=" + keywords);
        httpBackend.expectGET(request).respond(results)
      }

      return ctrl = $controller('MixesController', { 
        $scope: scope, 
        $location: location 
      });

    });
  };

  beforeEach(function() { 
    module("blueberry"); 
  });

  afterEach(function() {
    httpBackend.verifyNoOutstandingExpectation();
    httpBackend.verifyNoOutstandingRequest();
  });

  // Start the tests!

  describe('controller initialization', function() {
    describe('when no keywords are present', function() {
      beforeEach(setupController());  
      it('defaults to no recipes', function() {
        expect(ctrl.selectedMixes).toEqualData([]);
      });
    });

    describe('with keywords', function() {
      var keywords, mixList;

      keywords = 'foo';
      mixList  = [
        {
          id: 1,
          name: 'Supersonic Overdrive'
        },
        {
          id: 2,
          name: 'ChilZone'
        },
        {
          id: 3,
          name: 'Pot Pourri'
        }
      ];

      beforeEach(function() {
        setupController(keywords, mixList);
        return httpBackend.flush();
      })

      it('calls the back-end', function() {
        return expect(scope.mixList).toEqualData(mixList);
      });
    });

    describe('search()', function() {
      beforeEach(function() {
        setupController();
      });

      it('redirects to itself with a keyword param', function() {
        keywords = 'foo';
        scope.search(keywords);
        expect(location.path()).toBe('/');
        expect(location.search()).toEqualData({keywords: keywords});
      });
    });
  });
  

});